### PR TITLE
Remove the string to scalar estimator map from EstimatorManagerNew

### PR DESCRIPTION
**Pull request by PDoakORNL**
## Proposed changes

The legacy implementation of the EstimatorManager keeps a string to scalar estimator map.  This is only used to determine which estimator is the "main estimator" which would be called "LocalEnergy" in the map regardless of the name or type attribute for the estimator in the input.  This would allow the LocalEnegy estimator to be replaced by whichever had shown up in and `<estimator name="name_of_a_main_estimator>` node most recently.

As it turns out this was an overly complex solution that obfuscated the special states of the ScalarEstimatorBase subtype associated with a particular driver.

I have retained the ability to add miscellaneous scalar drivers that use the same scalar inputs strategy as the main estimator. Currently there doesn't seem to be an example of such except the Collectibles driver which is not part of the new design and which I won't explain here.

Also in the PR support is added to specify the main estimator in a sensible way, see #3645. for now it also supports the inconsistent legacy method.

If a type attribute is specified, then, as with almost every other instance of the "name" attribute, the user has free choice. If no type is specified then for "backward" compatibility the "name" attribute will be used for the type. 
If only type is specified the name will be atomatically set to match the type. Although as far as I can tell it is never used. And I only leave it as part of the acceptable input for the input classes to support future use of the name attribute and legacy input files.

If no main estimator is set it will as with the legacy code automatically use the LocalEnergyEstimator whether this is appropriate for a driver or not.

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- Removing unecessary and unclear code.
- Refactoring (no functional changes, minor and propagated API changes)
- Testing changes (e.g. new unit/integration/performance tests)
- Documentation changes

### Does this introduce a breaking change?
- No

## What systems has this change been tested on?
leconte

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- No. This PR is up to date with current the current state of 'develop' needs rebase after #3987 
- Yes Code added or changed in the PR has been clang-formatted
- Yes This PR adds tests to cover any new code, or to catch a bug that is being fixed
- Yes Documentation has been added (if appropriate)

### CODE REVIEW

File: src/QMCDrivers/QMCDriverNew.cpp [hunk 1]
```diff
@@ -466,22 +466,22 @@ QMCDriverNew::AdjustedWalkerCounts QMCDriverNew::adjustGlobalWalkerCount(int num
  */
 void QMCDriverNew::endBlock()
 {
-  ScopedTimer local_timer(timers_.endblock_timer);
```
Comment (markdewing) [inline src/QMCDrivers/QMCDriverNew.cpp:469]: Is removal of this timer intended?

Comment (PDoakORNL) [reply src/QMCDrivers/QMCDriverNew.cpp:469]: no that's a mistake with my merge.

File: src/Estimators/ScalarEstimatorBase.h [hunk 2]
```diff
@@ -61,6 +61,7 @@ struct ScalarEstimatorBase
 
   virtual ~ScalarEstimatorBase() {}
 
+  virtual bool isMainEstimator() { return false; }
```
Comment (ye-luo) [inline src/Estimators/ScalarEstimatorBase.h:64]: isMainEstimator() const

File: src/Estimators/ScalarEstimatorBase.h [hunk 3]
```diff
@@ -162,6 +163,8 @@ struct ScalarEstimatorBase
 
   ///clone the object
   virtual ScalarEstimatorBase* clone() = 0;
+
+  virtual const std::string& getSubTypeStr() = 0;
```
Comment (ye-luo) [inline src/Estimators/ScalarEstimatorBase.h:167]: getSubTypeStr() const

File: src/Estimators/ScalarEstimatorInputs.h [hunk 4]
```diff
@@ -45,19 +57,55 @@ class CSLocalEnergyInput
...
+  using Consumer = CSEnergyEstimator;
+  CSLocalEnergyInput() = default;
   CSLocalEnergyInput(xmlNodePtr cur);
   int get_n_psi() const { return n_psi_; }
-
+  const std::string& get_name() { return name_; }
+  const std::string& get_type() { return type_; }
 private:
+  std::string name_;
+  std::string type_;
```
Comment (ye-luo) [inline src/Estimators/ScalarEstimatorInputs.h:75]: Need full documentation of this class. What is a name, what is type.

File: src/Estimators/ScalarEstimatorInputs.h [hunk 5]
```diff
@@ -45,19 +57,55 @@ class CSLocalEnergyInput
...
   };
 
 public:
+  using Consumer = CSEnergyEstimator;
+  CSLocalEnergyInput() = default;
   CSLocalEnergyInput(xmlNodePtr cur);
   int get_n_psi() const { return n_psi_; }
-
+  const std::string& get_name() { return name_; }
+  const std::string& get_type() { return type_; }
```
Comment (ye-luo) [inline src/Estimators/ScalarEstimatorInputs.h:72]: both get function need const

**Review by ye-luo — COMMENTED** [review]
Could you take out
Estimators/ScalarEstimatorBase and derived classes.
ScalarEstimatorInputs.h into a separate PR?
It is like to reduce the complex of reviewing the manager side changes.

**Comment by PDoakORNL** [comment]
I was hoping to just get the changes to the scalar estimators in one PR. I think only isMainEstimator is necessary for the state purpose of this PR. 

I'm over budget on QMCPACK so I really shouldn't do that until next week.

**Comment by ye-luo** [comment]
> I was hoping to just get the changes to the scalar estimators in one PR. I think only isMainEstimator is necessary for the state purpose of this PR.
> 
> I'm over budget on QMCPACK so I really shouldn't do that until next week.

Then please address my review regarding the files I mentioned and then I might regain some energy to review the rest of code.

File: src/Estimators/LocalEnergyOnlyEstimator.h [hunk 6]
```diff
@@ -66,6 +66,9 @@ struct LocalEnergyOnlyEstimator : public ScalarEstimatorBase
   }
 
   LocalEnergyOnlyEstimator* clone() override { return new LocalEnergyOnlyEstimator(); }
+
+  std::string type_str = "LocalEnergyOnlyEstimatorNotSupportedInBatchedVersion";
```
Comment (ye-luo) [inline src/Estimators/LocalEnergyOnlyEstimator.h:70]: Also better to protect these names with const.

**Comment by PDoakORNL** [comment]
test this please

File: src/Estimators/EstimatorManagerCrowd.h [hunk 7]
```diff
@@ -86,8 +86,9 @@ class EstimatorManagerCrowd
   ///total weight accumulated in a block
   RealType block_weight_;
 
+  UPtr<ScalarEstimatorBase> main_estimator_;
   ///estimators of simple scalars
-  std::vector<std::unique_ptr<EstimatorType>> scalar_estimators_;
+  std::vector<std::unique_ptr<ScalarEstimatorBase>> scalar_estimators_;
 
   std::vector<std::unique_ptr<OperatorEstBase>> operator_ests_;
```
Comment (ye-luo) [inline src/Estimators/EstimatorManagerCrowd.h:93]: We will need to do some refactoring in a **separate** PR. main_estimator_, scalar_estimators_, operator_ests_ should be const unique_ptr as they are being settled in the constructor and never intended to be changed throughout its life time.

EstimatorManagerCrowd should not be a friend of EstimatorManagerNew and EstimatorManagerNew should have APIs to return clones of these collection of estimators.

File: src/Estimators/EstimatorManagerNew.cpp [hunk 8]
```diff
@@ -425,22 +443,17 @@ bool EstimatorManagerNew::put(QMCHamiltonian& H,
...
-    EstimatorMap[aname] = n;
-  }
-  else
-  {
-    n = (*it).second;
-    app_log() << "  EstimatorManagerNew::add replace " << aname << " estimator." << std::endl;
-    Estimators[n] = std::move(newestimator);
-  }
-  return n;
+  if(main_estimator_ != nullptr)
```
Comment (ye-luo) [inline src/Estimators/EstimatorManagerNew.cpp:448]: Can be short as `if(!main_estimator_)` but it is more of personal taste.

Comment (ye-luo) [inline src/Estimators/EstimatorManagerNew.cpp:448]: Using `empty()`?

File: src/Estimators/tests/EstimatorManagerNewTest.cpp [hunk 9]
```diff
@@ -92,12 +143,12 @@ std::vector<QMCTraits::RealType> EstimatorManagerNewTest::generateGoodOperatorDa
...
-void EstimatorManagerNewTest::collectScalarEstimators()
+void EstimatorManagerNewTest::collectMainEstimators()
 {
-  RefVector<ScalarEstimatorBase> est_list = makeRefVector<ScalarEstimatorBase>(estimators_);
-  em.collectScalarEstimators(est_list);
+  auto crowd_main_estimators = makeRefVector<ScalarEstimatorBase>(estimators_);
+  em.collectMainEstimators(crowd_main_estimators);
 }
-
+
```
Comment (ye-luo) [inline src/Estimators/tests/EstimatorManagerNewTest.cpp:151]: Formatting?

Comment (ye-luo) [inline src/Estimators/tests/EstimatorManagerNewTest.cpp:151]: Formatting?

**Review by ye-luo — APPROVED** [review]
LGTM

**Comment by ye-luo** [comment]
Test this please

**Pull request merged**